### ISSUE TITLE
byref-pointer assignment and --warnon:1182

### ISSUE THREAD

**Issue by xenocons** [Microsoft/visualfsharp#405]
Hello,

Once again I am hesitant to log this as an issue because it may be a matter of reading the F# spec, If this is a stackoverflow question please redirect me, but it smells like a bit of a bug to me (albeit very minor).

When using warnon:1182 (warning for unused variables), it seems assigning to a byref-pointer causes the compiler to think that the value is unsed.

Ex.1
```
    let (=@) (e:int byref) n = e <- n; n
``` 
Warning generated is "the value 'e' is unused"

This is also the case for the following, but in this case its valid as p is in the scope and not returned:
Ex.2
```
    let x() = 
        let mutable p = 1
        p <- 1
        4
```

However, the following is recognised by the compiler and a warning is not generated:
Ex.3
```
    let mutable p = 1
    let y() = 
        p <- 1
        4
```

I would expect the same warning behaviour as 'Ex.3' to be exhibited when compiling 'Ex.1'

Is this one of the mysterious reasons that the 1182 warning is disabled by default? Or merely expected?

As a fix it is possible to get rid of the warning in a cheapish way by returning the 'e' parameter, it gets dereferenced automagically by the F# compiler when used as a return value (????):

```
    let (=@) (e:int byref) n = 
        e <- n
        e
```

Or it may be possible to use ignore on 'e'.

Edit: updated to try to make a cleaner argument after more research.

**Comment by enricosada** [comment]
You can "ignore" a variabile using `_` prefix. 
For example `_e`

**Comment by latkin** [comment]
Seems like a bug to me, I would expect any read or write to qualify a value as "used."

**Comment by dsyme** [comment]
Has this been fixed in F# 4.0?  I see this extra line, which I believe is the fix: https://github.com/Microsoft/visualfsharp/blob/fsharp4/src/fsharp/tc.fs#L8200

**Comment by latkin** [comment]
Oh, yes this does appear to already be fixed in 4.0 :laughing:
